### ISSUE TITLE
ES output failing with long messages

### ISSUE THREAD

**Issue by gmoskovicz** [logstash-plugins/logstash-output-elasticsearch#576]
- Version: Elasticsearch 5.2.2
- Operating System: ANY
- Config File (if you have sensitive info, please remove it)

```
input {
    file {
        path => [ "/Users/Gabriel/Documents/ElasticSearch/logstash-5.2.2/file/content.txt" ]
        type => "test"
        start_position => "beginning"
        sincedb_path => "/dev/null"
        ignore_older => 0
    }
}
filter {

}
output {
    elasticsearch {
        hosts => [ "localhost:9200" ]
        index => "test-1"
    }
}
```

- Sample Data:

Content.txt file (unzip it after download): [content.txt.zip](https://github.com/logstash-plugins/logstash-output-elasticsearch/files/850986/content.txt.zip)

- Steps to Reproduce:

[1] Run the above config with `content.txt` file (>50MB 1 line file)
[2] Wait until the message is processed 
[3] Verify that you get the following error:

```
[2017-03-17T12:12:44,260][ERROR][logstash.outputs.elasticsearch] Got a bad response code from
 server, but this code is not considered retryable. Request will be dropped {:code=>400, 
:response_body=>"{\"error\":{\"root_cause\":[{\"type\":\"parse_exception\",\"reason\":\"Failed to
 derive xcontent\"}],\"type\":\"parse_exception\",\"reason\":\"Failed to derive 
xcontent\"},\"status\":400}"}
```

In wireshark you can verify that the operation is sent like this (not the Content Length 0):

```
Hypertext Transfer Protocol
    POST /_bulk HTTP/1.1\r\n
    Connection: Keep-Alive\r\n
    Content-Type: application/json\r\n
    Content-Length: 0\r\n
        [Content length: 0]
    Host: localhost:9200\r\n
    User-Agent: Manticore 0.6.1\r\n
    Accept-Encoding: gzip,deflate\r\n
    \r\n
    [Full request URI: http://localhost:9200/_bulk]
    [HTTP request 4/4]
    [Prev request in frame: 17]
    [Response in frame: 23]
```

Basically when you have a very big file, Logstash is failing to send it correctly as a bulk request.

**Comment by gmoskovicz** [comment]
CC @ajoliveira

**Comment by jsvd** [comment]
The problem seems to come from [this piece of code](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/http_client.rb#L106-L125):

```ruby
      body_stream = StringIO.new
      if http_compression
        body_stream.set_encoding "BINARY"
        stream_writer = Zlib::GzipWriter.new(body_stream, Zlib::DEFAULT_COMPRESSION, Zlib::DEFAULT_STRATEGY)
      else 
        stream_writer = body_stream
      end
      bulk_responses = []
      bulk_actions.each do |action|
        as_json = action.is_a?(Array) ?
                    action.map {|line| LogStash::Json.dump(line)}.join("\n") :
                    LogStash::Json.dump(action)
        as_json << "\n"
        bulk_responses << bulk_send(body_stream) if (body_stream.size + as_json.bytesize) > TARGET_BULK_BYTES
        stream_writer.write(as_json)
      end
```

Starting the bulk operation, if the first element is over the TARGET_BULK_BYTES we flush the "current buffer". However in the scenario the buffer is empty, which creates the exception we see.

Adding an extra check to only flush if both:
1. buffer isn't empty
2. buffer + current element size > TARGET_BULK_BYTES

Will fix the issue.

**Comment by jsvd** [comment]
created https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/577 to fix this issue

**Comment by PhaedrusTheGreek** [comment]
Just for correlation - I can reproduce the same issue in Logstash 5.6.9, but there's a different error message:

```
[2018-04-30T10:40:16,767][ERROR][logstash.outputs.elasticsearch] Encountered a retryable error. Will Retry with exponential backoff  {:code=>400, :url=>"http://localhost:9200/_bulk", :body=>"{\"error\":{\"root_cause\":[{\"type\":\"parse_exception\",\"reason\":\"request body is required\"}],\"type\":\"parse_exception\",\"reason\":\"request body is required\"},\"status\":400}"}
```

**Comment by jsvd** [comment]
this was solved by #577 in the plugin versions that ship with logstash 6.x, https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/737 is the backport for versions shipped with logstash 5.x
